// node-hue-api/dist/esm/api/Rules.js
import { ApiDefinition } from './http/ApiDefinition';
import { rulesApi } from './http/endpoints/rules';
export class Rules extends ApiDefinition {
    constructor(hueApi) {
        super(hueApi);
    }
    getAll() {
        return this.execute(rulesApi.getAll);
    }
    getRule(id) {
        return this.execute(rulesApi.getRule, { id: id });
    }
    // /**
    //  * @deprecated since 4.x, use getRule(id) instead
    //  */
    // get(id) {
    //   util.deprecatedFunction('5.x', 'rules.get(id)', 'Use rules.getRule(id) instead.');
    //   return this.getRule(id);
    // }
    // getOrphanedRules() {
    //   return this._filterRules(rule => {return rule.status === 'resourcedeleted'});
    // }
    //
    // getDisabledRules() {
    //   return this._filterRules(rule => {return rule.status === 'disabled'});
    // }
    getRuleByName(name) {
        return this.getAll()
            .then(rules => {
            return rules.filter(rule => rule.name === name);
        });
    }
    createRule(rule) {
        const self = this;
        return self.execute(rulesApi.createRule, { rule: rule })
            .then(data => {
            return self.getRule(data.id);
        });
    }
    deleteRule(id) {
        return this.execute(rulesApi.deleteRule, { id: id });
    }
    updateRule(rule) {
        return this.execute(rulesApi.updateRule, { id: rule.id, rule: rule });
    }
}
