// node-hue-api/dist/cjs/lib/api/ResourceLinks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceLinks = void 0;
const ApiDefinition_1 = require("./http/ApiDefinition");
const resourceLinks_1 = require("./http/endpoints/resourceLinks");
class ResourceLinks extends ApiDefinition_1.ApiDefinition {
    constructor(hueApi) {
        super(hueApi);
    }
    getAll() {
        return this.execute(resourceLinks_1.resourceLinksApi.getAll);
    }
    getResourceLink(id) {
        return this.execute(resourceLinks_1.resourceLinksApi.getResourceLink, { id: id });
    }
    getResourceLinkByName(name) {
        return this.getAll()
            .then(resourceLinks => {
            return resourceLinks.filter(resourceLink => resourceLink.name === name);
        });
    }
    createResourceLink(resourceLink) {
        const self = this;
        return self.execute(resourceLinks_1.resourceLinksApi.createResourceLink, { resourceLink: resourceLink })
            .then((result) => {
            return self.getResourceLink(result.id);
        });
    }
    deleteResourceLink(id) {
        return this.execute(resourceLinks_1.resourceLinksApi.deleteResourceLink, { id: id });
    }
    updateResourceLink(resourceLink) {
        return this.execute(resourceLinks_1.resourceLinksApi.updateResourceLink, { id: resourceLink.id, resourceLink: resourceLink });
    }
}
exports.ResourceLinks = ResourceLinks;
