// @peter-murray/hue-bridge-model/dist/cjs/model/Light.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Light = void 0;
const BridgeObjectWithId_1 = require("./BridgeObjectWithId");
const ColorGamuts_1 = require("./ColorGamuts");
const types_1 = require("../types");
const MODEL_TO_COLOR_GAMUT = {
    'LCT001': 'B',
    'LCT007': 'B',
    'LCT010': 'C',
    'LCT014': 'C',
    'LCT015': 'C',
    'LCT016': 'C',
    'LCT002': 'B',
    'LCT003': 'B',
    'LCT011': 'C',
    // 'LCT024': 'C', //TODO this can be read from capabilities.control.colorgamut and colorgamuttype now - REMOVE THIS
    'LTW011': '2200K-6500K',
    'LST001': 'A',
    'LLC010': 'A',
    'LLC011': 'A',
    'LLC012': 'A',
    'LLC006': 'A',
    'LLC005': 'A',
    'LLC007': 'A',
    'LLC014': 'A',
    'LLC013': 'A',
    'LLM001': 'B',
    'LLM010': '2200K-6500K',
    'LLM011': '2200K-6500K',
    'LTW001': '2200K-6500K',
    'LTW004': '2200K-6500K',
    'LTW010': '2200K-6500K',
    'LTW015': '2200K-6500K',
    'LTW013': '2200K-6500K',
    'LTW014': '2200K-6500K',
    'LLC020': 'C',
    'LST002': 'C',
    'LCT012': 'C',
    'LTW012': '2200K-6500K',
    // Lamps
    'LTP001': '2200K-6500K',
    'LTP002': '2200K-6500K',
    'LTP003': '2200K-6500K',
    'LTP004': '2200K-6500K',
    'LTP005': '2200K-6500K',
    'LTF001': '2200K-6500K',
    'LTF002': '2200K-6500K',
    'LTC001': '2200K-6500K',
    'LTC002': '2200K-6500K',
    'LTC003': '2200K-6500K',
    'LTC004': '2200K-6500K',
    'LTC011': '2200K-6500K',
    'LTC012': '2200K-6500K',
    'LTD001': '2200K-6500K',
    'LTD002': '2200K-6500K',
    'LFF001': '2200K-6500K',
    'LTT001': '2200K-6500K',
    'LDT001': '2200K-6500K',
};
const ATTRIBUTES = [
    new types_1.UInt8Type({ name: 'id' }),
    new types_1.StringType({ name: 'name', minLength: 0, maxLength: 32 }),
    new types_1.StringType({ name: 'type' }),
    new types_1.StringType({ name: 'modelid' }),
    new types_1.StringType({ name: 'manufacturername' }),
    new types_1.StringType({ name: 'uniqueid' }),
    new types_1.StringType({ name: 'productname' }),
    new types_1.StringType({ name: 'productid' }),
    new types_1.ObjectType({ name: 'state' }),
    new types_1.ObjectType({ name: 'capabilities' }),
    new types_1.ObjectType({ name: 'config' }),
    new types_1.ObjectType({
        name: 'swupdate',
        types: [
            new types_1.StringType({ name: 'state' }),
            new types_1.StringType({ name: 'lastinstall' }),
        ]
    }),
    new types_1.StringType({ name: 'swversion' }),
    new types_1.StringType({ name: 'swconfigid' }),
];
//TODO add support for making it easier to set power failure modes config.startup.mode = 'powerfail'
class Light extends BridgeObjectWithId_1.BridgeObjectWithId {
    constructor(id) {
        super(ATTRIBUTES, id);
        this.mappedColorGamut = null;
    }
    get name() {
        return this.getAttributeValue('name');
    }
    set name(value) {
        this.setAttributeValue('name', value);
    }
    get type() {
        return this.getAttributeValue('type');
    }
    get modelid() {
        return this.getAttributeValue('modelid');
    }
    get manufacturername() {
        return this.getAttributeValue('manufacturername');
    }
    get uniqueid() {
        return this.getAttributeValue('uniqueid');
    }
    get productid() {
        return this.getAttributeValue('productid');
    }
    get productname() {
        return this.getAttributeValue('productname');
    }
    get swversion() {
        return this.getAttributeValue('swversion');
    }
    get swupdate() {
        return this.getAttributeValue('swupdate');
    }
    get state() {
        return this.getAttributeValue('state');
    }
    get capabilities() {
        return this.getAttributeValue('capabilities');
    }
    get colorGamut() {
        if (this.mappedColorGamut && this.mappedColorGamut !== '2200K-6500K') {
            return (0, ColorGamuts_1.getColorGamut)(this.mappedColorGamut);
        }
        else {
            return null;
        }
    }
    /**
     * Gets the supported states that the light will accept.
     */
    getSupportedStates() {
        const states = Object.keys(this.state);
        // transitiontime is no longer provided in the light state raw data values from the Hue API
        states.push('transitiontime');
        // If there is a corresponding settings, then include the xxx_inc variant
        ['bri', 'sat', 'hue', 'ct', 'xy'].forEach(key => {
            if (states.indexOf(key) > -1) {
                states.push(`${key}_inc`);
            }
        });
        return states;
    }
    _populate(data) {
        if (data) {
            this.mappedColorGamut = resolveColorGamutFromData(data);
        }
        else {
            this.mappedColorGamut = null;
        }
        return super._populate(data);
    }
}
exports.Light = Light;
function getValueforKey(key, data) {
    // Use dot notation to get nested values
    const path = key.split('.');
    let target = data, value = null;
    path.forEach(part => {
        if (target != null) {
            // @ts-ignore
            value = target[part];
            target = value;
        }
        else {
            target = null;
        }
    });
    return `${value}`;
}
function resolveColorGamutFromData(data) {
    // Newer Hue devices report their own color gamuts under 'capabilities.control.colorgamuttype'
    let colorGamutType = getValueforKey('capabilities.control.colorgamuttype', data);
    if (!colorGamutType) {
        colorGamutType = MODEL_TO_COLOR_GAMUT[data.modelid];
    }
    return colorGamutType;
}
