// node-hue-api/dist/cjs/lib/api/http/sslCertificate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSSLCertificate = void 0;
const https_1 = __importDefault(require("https"));
function getSSLCertificate(url, timeout) {
    const options = getRequestOptions(url, timeout);
    return performRequest(options);
}
exports.getSSLCertificate = getSSLCertificate;
function performRequest(options) {
    return new Promise((resolve, reject) => {
        https_1.default.get(options, (res) => {
            // @ts-ignore
            const cert = res.socket.getPeerCertificate();
            if (!cert || Object.keys(cert).length < 1) {
                reject(`Bridge did not supply a certificate`);
            }
            else {
                resolve(cert);
            }
        });
    });
}
function getRequestOptions(url, timeout) {
    return {
        hostname: url,
        agent: false,
        rejectUnauthorized: false,
        ciphers: 'ALL'
    };
}
